### PR TITLE
Loading context data into TOAST

### PR DESCRIPTION
**Pull request by tskisner**
General interface for parallel loading of context data into a toast observation.  Does not yet include hook for running pre-processing on load.  Still missing a roundtrip unit test.

### CODE REVIEW

File: sotodlib/toast/ops/load_context.py [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+            name_col = Column(name="name", data=det_props["det_info_readout_id"])
+
+            if "focal_plane_xi" in det_props.colnames:
+                quat_data = toast.instrument_coords.xieta_to_quat(
+                    det_props["focal_plane_xi"],
+                    det_props["focal_plane_eta"],
+                    det_props["focal_plane_gamma"],
+                )
+            else:
+                # No detector offsets yet
```
Comment (keskitalo) [inline sotodlib/toast/ops/load_context.py:338]: I'm a little bit conflicted with defaulting to unit quaternions here. Perhaps we should require the user to indicate if they want detector quaternions or not and then we'll know if their absence is an error?

**Review by keskitalo — APPROVED** [review]
Looks great! Only minor comments. Can't wait to start analyzing ACT data.

File: sotodlib/toast/ops/sim_source.py
[the remark this answers is not in the document]
Comment (tskisner) [reply sotodlib/toast/ops/sim_source.py:1]: These were just cleaning up unused imports and empty lines, but I agree we should isolate those.  Maybe time for a cleanup PR that runs black / isort on the toast submodule.

File: tests/test_toast_context.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+class ToastContextTest(TestCase):
+    def setUp(self):
+        fixture_name = os.path.splitext(os.path.basename(__file__))[0]
+        if not toast_available:
+            print("TOAST cannot be imported- skipping unit tests", flush=True)
+            return
+        world, procs, rank = toast.get_world()
+        self.outdir = create_outdir(fixture_name, mpicomm=world)
+
+    def test_context_load(self):
```
[the remark this answers is not in the document]
Comment (tskisner) [reply tests/test_toast_context.py:48]: As previously discussed, in the interest of progress, a fancier unit test here can wait.  I have run this branch in jupyter notebooks to load both ACT data and satp1 data from the site.  Once there is an easier way to simulate data on disk, create a context, load it back in and verify, then we can add that here.  Actually pieces of that code exist and were removed from this PR in order to not hold it up.

File: sotodlib/toast/io/save_book.py [hunk 3]
```diff
@@ -629,27 +640,45 @@ def write_book(
...
+    # frame.  NOTE:  we copy what is done in the BookBinder class.
     if comm.group_rank == 0:
         obs_meta = dict()
         obs_meta["obs_id"] = book_name
+        obs_meta["type"] = "obs"
+        obs_meta["book_id"] = book_name
         obs_meta["observatory"] = "Simons Observatory"
         obs_meta["telescope"] = obs_props["telescope_name"]
         obs_meta["stream_ids"] = [str(x) for x in all_wafers.keys()]
+        # FIXME: is this right?
```
[the remark this answers is not in the document]
Comment (tskisner) [reply sotodlib/toast/io/save_book.py:652]: This was more of a stale note-to-self, but I can go look at that other code and confirm.

[the remark this answers is not in the document]
Comment (tskisner) [reply sotodlib/toast/io/save_book.py:652]: Looks like these corresponding values in the bookbinder / imprinter are just using timestamps from the original frames.  So what we are doing here is consistent with that for the simulation case.  I'll remove the FIXME

File: sotodlib/toast/ops/load_context.py (hunk 1 above)
Comment (tskisner) [reply sotodlib/toast/ops/load_context.py:338]: Only some toast operators require the detector pointing offsets.  If the user was loading data without those offsets, then presumably they would be intending to either not use them or to run code to estimate / populate them.  In either case, we could simply exclude them from the focalplane table.  Then any attempted use would raise an exception.  I'll make that change.

Comment (tskisner) [reply sotodlib/toast/ops/load_context.py:339]: On further inspection, this change would require a small upstream in the `toast.Focalplane` class.  I'm going to leave the null quats here for now, but will fix after that change upstream.

**Pull request merged**